internal/attestation/tdx/validator.go

```diff
@@ -132,3 +157,69 @@ func (t tdxReport) HostData() []byte {
 func (t tdxReport) ClaimsToCertExtension() ([]pkix.Extension, error) {
 	return claimsToCertExtension(t.quote)
 }
+
+// getPIID extracts the PIID from the PCK certificate inside a TDX quote.
+func getPIID(quote *tdx.QuoteV4) ([]byte, error) {
+	pckCertChain := quote.GetSignedData().GetCertificationData().GetQeReportCertificationData().GetPckCertificateChainData().PckCertChain
+
+	// The certChain input is a concatenated list of PEM-encoded X.509 certificates.
+	// https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_TDX_DCAP_Quoting_Library_API.pdf, A.3.9
+
+	var pckBlock *pem.Block
+	var pck *x509.Certificate
+	for len(pckCertChain) > 0 {
+		pckBlock, pckCertChain = pem.Decode(pckCertChain)
+		if pckBlock == nil {
+			break
+		}
+		candidate, err := x509.ParseCertificate(pckBlock.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("parsing PCK certificate: %w", err)
+		}
+
+		if candidate.Subject.CommonName == "Intel SGX PCK Certificate" {
+			pck = candidate
+			break
+		}
+	}
+	if pck == nil {
+		return nil, fmt.Errorf("no PCK certificate found in TDX quote")
+	}
+
+	var sgxExtensions cryptobyte.String
+	for _, ext := range pck.Extensions {
+		if !ext.Id.Equal(oid.SGXExtensionsOID) {
+			continue
+		}
+		extValue := cryptobyte.String(ext.Value)
+		if !extValue.ReadASN1(&sgxExtensions, cryptobyte_asn1.SEQUENCE) {
+			return nil, fmt.Errorf("could not read SGX extensions from PCK cert")
+		}
+	}
+	if sgxExtensions == nil {
+		return nil, fmt.Errorf("no SGX extensions found on PCK certificate")
+	}
+	for !sgxExtensions.Empty() {
+		var extension cryptobyte.String
+		if !sgxExtensions.ReadASN1(&extension, cryptobyte_asn1.SEQUENCE) {
+			return nil, fmt.Errorf("could not parse SGX extension")
+		}
+		var id asn1.ObjectIdentifier
+		if !extension.ReadASN1ObjectIdentifier(&id) {
+			return nil, fmt.Errorf("could not parse SGX extension OID")
+		}
+		if !id.Equal(oid.PlatformInstanceIDOID) {
+			continue
+		}
+		var piid cryptobyte.String
+		if !extension.ReadASN1(&piid, cryptobyte_asn1.OCTET_STRING) {
+			return nil, fmt.Errorf("could not parse SGX extension value")
+		}
+
+		if len(piid) != 16 {
+			return nil, fmt.Errorf("expected PIID of size 16, got %d", len(piid))
+		}
+		return piid, nil
+	}
+	return nil, fmt.Errorf("no PIID extension found in PCK SGX extensions")
+}
```

Comment:
Is this something we could try to upstream to go-tdx-guest?

Reply:
Mentioned the upstream API and my intent to contribute that in a comment.